### ISSUE TITLE
Expose line number of the definition of a JavaMember

### ISSUE THREAD

**Issue by rweisleder** [TNG/ArchUnit#75]
For a custom ArchCondition that checks the definition of a JavaClass/JavaMethod/JavaField it would be useful to append the declaration line to the error message as clickable link.

Example:
```java
private ArchCondition<JavaMethod> haveNameStartingWithLowerCase() {
    return new ArchCondition<JavaMethod>("have name starting with lower case") {
        @Override
        public void check(JavaMethod item, ConditionEvents events) {
            boolean satisfied = Character.isLowerCase(item.getName().charAt(0));
            String message = item.getFullName() + " " + Formatters.formatLocation(item.getOwner(), item.getDeclarationLineNumber());
            events.add(new SimpleConditionEvent(item, satisfied, message));
        }
    };
}
```
would print
```
java.lang.AssertionError: Architecture Violation [Priority: MEDIUM] - Rule 'methods should have name starting with lower case' was violated (1 times):
demo.GreetEndpoint.greet(java.lang.String) (GreetEndpoint.java:23)
```

Relates to #38

**Comment by codecholeric** [comment]
I know, this issue came up in the past. Surprisingly, there doesn't seem to be a clean way to determine the line number from Java bytecode (I would already have added it otherwise).
Every solution I found (I looked at Javassist, etc.) at best had some workaround.
The problem is, that there is no trace of the line number of a member within the bytecode as far as I know, so the only way is, to look at the first statement of a method, and do something shady like "subtract 1" to get to the declaration (for fields there is not even such a workaround AFAIK)
I'm open to suggestions though, if someone knows a solution to this!! But I think it just is an intrinsic limitation of Java bytecode.
Also compare: https://stackoverflow.com/questions/49986872/how-to-determine-line-number-for-the-method-with-java-asm

**Comment by bedla** [comment]
My poor-man's solution is
```
        final int lineNumber = Stream.of(javaCodeUnit.getMethodCallsFromSelf(),
                javaCodeUnit.getConstructorCallsFromSelf(),
                javaCodeUnit.getFieldAccesses())
                .flatMap(Collection::stream)
                .mapToInt(JavaAccess::getLineNumber)
                .min()
                .orElse(0);
```

**Comment by codecholeric** [comment]
Yeah, for methods that would work, you could subtract 1, hope that nobody includes a comment or empty line, and it would match the method declaration in most cases :wink: 
However, I don't see any way to do this for field declarations, so it would really only address non-empty methods and constructors in a heuristic way.
If we don't find any better solution, we might add a `guessLineNumber(JavaCodeUnit)` method with a respective JavaDoc, and keep it 0 for field declarations.
